### DISCUSSION TITLE
Use Modbus Bridge and Smart Meter Interface simultaneously ?

### DISCUSSION THREAD

**Discussion by erikgieseler** [arendst/Tasmota#19860] [Q&A] [answered]
Hello, 

is it possible to use Modbus Bridge and Smart Meter Interface at the same time with different RS485 to TTL adapters ?

The adapters are connected to the following GPIOs:
SML = GPIO13 and 27
Modbus Bridge = GPIO32 and 33

The SML should be used to read an energy meter automatically and the Modbus Bridge should be used for Node-Red.

When I start the SML script to read the energy meter, the TX LEDs of both RS485 to TTL adapters light up. Is this correct ?

Thanks in advance for the answers !

**Comment by gemu2015** [comment]
i assume you have a single smart meter and connect 2 RS485 adapters. this will not work.  you may only have one master and several slaves, not 2 masters on one slave on MODBUS

**Reply by erikgieseler** [reply]
The setup would look like this.
But theoretically these are 2 different RS485 buses or not?

![RS485](https://github.com/arendst/Tasmota/assets/72557610/f6626d93-6ee9-47f7-90f9-7fd92dba028f)

**Comment by gemu2015** [comment]
Yes this should work, you have 2 different smartmeters but why do both adapters blink when only sml driver automatically scans ?

**Reply by erikgieseler** [reply]
The two blinking LEDs also seemed strange to me. I can't say why that is.

**Comment by Noschvie** [comment]
Why do you need Modbus for Node-Red? Use MQTT instead.

**Reply by erikgieseler** [reply]
I need the second adapter as a Modbus bridge for a device that has only a RS485 port and no ethernet.
See also: https://tasmota.github.io/docs/Modbus-Bridge/

**Reply by Noschvie** [reply]
`device that has only a RS485 port and no ethernet` and running Node-Red, ok.

**Comment by pkkrusty** [comment]
Would you be better off defining the SML meter on hardware serial (16/17) and Modbus bridge on software? Or the other way around? What do you use in nodered to communicate with a Tasmota device using Modbus bridge?

**Reply by barbudor** [reply]
There is no dedicated uart pins on ESP32
Any gpio can be multiplexes to any of the 3 uarts

**Reply by erikgieseler** [reply]
I use the normal node-red-contrib-modbus package. Node-Red is not the problem.
The problem is that the TX LEDs of both TTL to RS485 adapters are blinking when only the SML script is running and nothing is running via Nore-Red. Normally only the TX LED of the SML adapter should blink.

**Comment by Jason2866** [comment]
Try to avoid GPIO 16/17 on the ESP32 if other GPIOs are free. There is GPIOs "blinking" when booting.
On this GPIOs is PSRAM connected. Tasmota does check for PSRAM and enables when found. This test does generate the GPIO blinking

**Comment by erikgieseler** [comment, marked as answer]
I have found the solution !

You have to tell the SML driver which UART to use in this case. 
You have to put `1,=so7,1` in the definition so that e.g. UART 1 should be used. 
The whole SML Script looks like this:

```
>D
>B
=>sensor53 r
>M 1
+1,27,m,0,9600,EM,13,10,03030018,03030034
1,=so7,1
1,030304ffffffff@i0:0.001,L3F1946 Leistung gesamt,W,load,0
1,030304ffffffff@i1:1,L3F1946 Bezug gesamt,kWh,bezug,2
#
```

Source: 
https://tasmota.github.io/docs/Smart-Meter-Interface/#special-commands
